<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Beluga · Beluga.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">Beluga.jl</span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Beluga</a><ul class="internal"><li><a class="tocitem" href="#Data-preparation-1"><span>Data preparation</span></a></li><li><a class="tocitem" href="#Reversible-jump-MCMC-for-the-inference-of-WGDs-in-a-phylogeny-1"><span>Reversible-jump MCMC for the inference of WGDs in a phylogeny</span></a></li></ul></li><li><a class="tocitem" href="api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Beluga</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Beluga</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/arzwa/Beluga.jl/blob/master/docs/src/index.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Beluga.jl-1"><a class="docs-heading-anchor" href="#Beluga.jl-1">Beluga.jl</a><a class="docs-heading-anchor-permalink" href="#Beluga.jl-1" title="Permalink"></a></h1><p>Beluga is a julia library for the statistical analysis of gene family evolution using phylogenetic birth-death processes. It&#39;s somewhat related to <a href="https://arzwa.github.io/Whale.jl/dev/index.html">Whale.jl</a> as it implements models of duplication loss and whole-genome duplication, but employs gene count data instead of gene trees. The library implements the MCMC sampler of Zwaenepoel &amp; Van de Peer (2019) as well as the reversible-jump MCMC sampler of Zwaenepoel &amp; Van de Peer (2020, <em>in preparation</em>).</p><h2 id="Data-preparation-1"><a class="docs-heading-anchor" href="#Data-preparation-1">Data preparation</a><a class="docs-heading-anchor-permalink" href="#Data-preparation-1" title="Permalink"></a></h2><p>To perform analyses with Beluga, you will need  </p><ol><li>An ultrametric species tree (time tree)</li><li>A <em>phylogenetic profile matrix</em>. If you have a bunch of protein fasta files for a set of species of interest, this can be easily obtained using e.g. <a href="https://github.com/davidemms/OrthoFinder">OrthoFinder</a>.</li></ol><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>If you only need the phylogenetic profile matrix from OrthoFinder, be sure to use the <code>-og</code> flag to stop the OrthoFinder pipeline after orthogroup inference. The phylogenetic profile matrix can be found in the <code>Orthogroups.GeneCount.tsv</code> file.</p></div></div><h2 id="Reversible-jump-MCMC-for-the-inference-of-WGDs-in-a-phylogeny-1"><a class="docs-heading-anchor" href="#Reversible-jump-MCMC-for-the-inference-of-WGDs-in-a-phylogeny-1">Reversible-jump MCMC for the inference of WGDs in a phylogeny</a><a class="docs-heading-anchor-permalink" href="#Reversible-jump-MCMC-for-the-inference-of-WGDs-in-a-phylogeny-1" title="Permalink"></a></h2><p>Load Beluga and required packages:</p><pre><code class="language-julia">using Beluga, CSV, Distributions</code></pre><p>Or, if you are running a julia session with multiple processes (when you started julia with <code>-p</code> option, or manually added workers using <code>addprocs</code>, see <a href="https://docs.julialang.org/en/v1/manual/parallel-computing/index.html#Parallel-Computing-1">julia docs</a>), run</p><pre><code class="language-julia">using CSV, Distributions
@everywhere using Beluga  # if julia is running in a parallel environment</code></pre><p>Then get some data (for instance from the <code>example</code> directory of the git repo)</p><pre><code class="language-julia">nw = readline(&quot;../../example/dicots/dicots.nw&quot;)
df = CSV.read(&quot;../../example/dicots/dicots-f01-25.csv&quot;)
model, data = DLWGD(nw, df, 1.0, 1.2, 0.9)</code></pre><pre><code class="language-none">(model = DLWGD{Float64,Beluga.TreeNode{Beluga.Branch{Float64}}}(17), data = PArray{Float64}(25))</code></pre><p><code>model</code> now refers to the duplication-loss and WGD model (with no WGDs for now), <code>data</code> refers to the phylogenetic profile matrix. The model was initialized with all duplication and loss rates set to 1 and 1.2 respectively. You can check this easily:</p><pre><code class="language-julia">getrates(model)</code></pre><pre><code class="language-none">2×17 Array{Float64,2}:
 1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  …  1.0  1.0  1.0  1.0  1.0  1.0  1.0
 1.2  1.2  1.2  1.2  1.2  1.2  1.2  1.2     1.2  1.2  1.2  1.2  1.2  1.2  1.2</code></pre><p>or to get the full parameter vector:</p><pre><code class="language-julia">asvector(model)</code></pre><pre><code class="language-none">35-element Array{Float64,1}:
 1.0
 1.0
 1.0
 1.0
 1.0
 1.0
 1.0
 1.0
 1.0
 1.0
 ⋮  
 1.2
 1.2
 1.2
 1.2
 1.2
 1.2
 1.2
 1.2
 0.9</code></pre><p>and you can easily compute likelihoods (and gradients)</p><pre><code class="language-julia">logpdf!(model, data)</code></pre><pre><code class="language-none">-251.03583571055535</code></pre><p>Now we proceed to the hierarchical model. There is no DSL available (à la Turing.jl, Mamba.jl or Soss.jl) but we use a fairly flexible prior struct. Here is an exaple for the (recommended) independent rates (IR) prior:</p><pre><code class="language-julia">prior = IRRevJumpPrior(
    # prior covariance matrix (Inverse Wishart prior)
    Σ₀=[1 0. ; 0. 1],

    # multivariate prior on the mean duplication and loss rates
    X₀=MvNormal([0., 0.], [1 0. ; 0. 1]),

    # prior on the number of WGDs (can be any discrete distribution)
    πK=DiscreteUniform(0,20),

    # prior on the WGD retention rate (should be `Beta`)
    πq=Beta(1,1),

    # prior on η
    πη=Beta(3,1),

    # tree length (determines prior on WGD times)
    Tl=treelength(model))</code></pre><pre><code class="language-none">IRRevJumpPrior{Distributions.MvNormal{Float64,PDMats.PDMat{Float64,Array{Float64,2}},Array{Float64,1}},Distributions.Beta{Float64},Distributions.Beta{Float64},Distributions.DiscreteUniform,Nothing}
  Σ₀: Array{Float64}((2, 2)) [1.0 0.0; 0.0 1.0]
  X₀: Distributions.MvNormal{Float64,PDMats.PDMat{Float64,Array{Float64,2}},Array{Float64,1}}
  πη: Distributions.Beta{Float64}
  πq: Distributions.Beta{Float64}
  πK: Distributions.DiscreteUniform
  πE: Nothing nothing
  Tl: Float64 0.8565633076700001
</code></pre><p>We can then construct a chain and run it</p><pre><code class="language-julia">chain = RevJumpChain(data=data, model=model, prior=prior)
init!(chain)
rjmcmc!(chain, 100, show=10)</code></pre><pre><code class="language-none">|    gen, pjmp,  k,k2,k3,    λ1,    λ2,    λ3,    μ1,    μ2,    μ3 …       logp,    logπ,     η
|     10, 0.00,  0, 0, 0, 0.928, 0.746, 1.192, 1.200, 0.896, 0.939 ⋯   -225.080, -26.440, 0.910
|     20, 0.10,  2, 0, 0, 1.065, 0.962, 1.137, 1.364, 1.396, 1.147 ⋯   -195.944, -38.145, 0.887
|     30, 0.17,  5, 0, 0, 0.742, 1.071, 1.302, 0.950, 1.554, 0.678 ⋯   -199.452, -34.051, 0.967
|     40, 0.17,  3, 0, 0, 0.681, 0.965, 0.793, 0.482, 0.366, 0.379 ⋯   -191.001, -34.089, 0.927
|     50, 0.18,  5, 0, 0, 1.126, 3.288, 1.343, 0.514, 0.526, 0.222 ⋯   -183.186, -40.310, 0.901
|     60, 0.22,  5, 0, 0, 1.465, 2.087, 0.840, 0.809, 1.232, 0.644 ⋯   -181.933, -39.380, 0.920
|     70, 0.24,  3, 0, 0, 1.543, 1.463, 1.634, 0.572, 0.535, 0.785 ⋯   -181.874, -41.485, 0.954
|     80, 0.24,  1, 0, 0, 1.615, 1.921, 0.495, 0.529, 1.048, 0.657 ⋯   -180.027, -49.636, 0.847
|     90, 0.23,  1, 0, 0, 1.372, 0.411, 2.126, 0.552, 1.312, 0.688 ⋯   -187.549, -47.028, 0.971
|    100, 0.21,  1, 0, 0, 1.289, 1.329, 1.717, 0.572, 0.603, 0.673 ⋯   -183.217, -43.741, 0.869</code></pre><p>This will log a part of the trace to stdout every <code>show</code> iterations, so that we&#39;re able to monitor a bit whether everything looks sensible. Of course in reality you would sample way longer than n=100 iterations.</p><p>Now we can analyze the trace (in <code>chain.trace</code>), write it to a file, etc. We can also compute Bayes factors to get an idea of the number of WGDs for each branch in the species tree.</p><pre><code class="language-julia">bayesfactors(chain, burnin=10)</code></pre><table class="data-frame"><thead><tr><th></th><th>branch</th><th>clade</th><th>k</th><th>eval1</th><th>eval2</th><th>p1</th><th>p0</th><th>π1</th><th>π0</th><th>t</th><th>tfrac</th><th>p2</th><th>π2</th><th>K1</th><th>K2</th><th>log10K1</th><th>log10K2</th></tr><tr><th></th><th>Int64</th><th>Array…</th><th>Int64</th><th>String</th><th>String</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th></tr></thead><tbody><p>27 rows × 17 columns</p><tr><th>1</th><td>2</td><td>[:ptr, :mtr, :ath, :cpa, :vvi]</td><td>0</td><td></td><td></td><td>1.0</td><td>NaN</td><td>0.922058</td><td>NaN</td><td>0.00702839</td><td>0.00820534</td><td>1.0</td><td>1.0</td><td>NaN</td><td>NaN</td><td>NaN</td><td>NaN</td></tr><tr><th>2</th><td>3</td><td>[:vvi]</td><td>0</td><td></td><td></td><td>1.0</td><td>NaN</td><td>0.350124</td><td>NaN</td><td>0.11</td><td>0.12842</td><td>1.0</td><td>1.0</td><td>NaN</td><td>NaN</td><td>NaN</td><td>NaN</td></tr><tr><th>3</th><td>4</td><td>[:ptr, :mtr, :ath, :cpa]</td><td>0</td><td></td><td></td><td>1.0</td><td>NaN</td><td>0.954654</td><td>NaN</td><td>0.004</td><td>0.00466982</td><td>1.0</td><td>1.0</td><td>NaN</td><td>NaN</td><td>NaN</td><td>NaN</td></tr><tr><th>4</th><td>5</td><td>[:ath, :cpa]</td><td>0</td><td></td><td></td><td>1.0</td><td>NaN</td><td>0.660827</td><td>NaN</td><td>0.0378</td><td>0.0441298</td><td>1.0</td><td>1.0</td><td>NaN</td><td>NaN</td><td>NaN</td><td>NaN</td></tr><tr><th>5</th><td>6</td><td>[:ath]</td><td>0</td><td></td><td></td><td>0.824176</td><td>NaN</td><td>0.493348</td><td>NaN</td><td>0.0682</td><td>0.0796205</td><td>1.0</td><td>1.0</td><td>NaN</td><td>NaN</td><td>NaN</td><td>NaN</td></tr><tr><th>6</th><td>6</td><td>[:ath]</td><td>1</td><td></td><td></td><td>0.175824</td><td>0.824176</td><td>0.303091</td><td>0.493348</td><td>0.0682</td><td>0.0796205</td><td>0.175824</td><td>0.506652</td><td>0.347247</td><td>0.207731</td><td>-0.459362</td><td>-0.682498</td></tr><tr><th>7</th><td>7</td><td>[:cpa]</td><td>0</td><td></td><td></td><td>0.791209</td><td>NaN</td><td>0.493348</td><td>NaN</td><td>0.0682</td><td>0.0796205</td><td>1.0</td><td>1.0</td><td>NaN</td><td>NaN</td><td>NaN</td><td>NaN</td></tr><tr><th>8</th><td>7</td><td>[:cpa]</td><td>1</td><td></td><td></td><td>0.186813</td><td>0.791209</td><td>0.303091</td><td>0.493348</td><td>0.0682</td><td>0.0796205</td><td>0.208791</td><td>0.506652</td><td>0.384323</td><td>0.256959</td><td>-0.415304</td><td>-0.590136</td></tr><tr><th>9</th><td>7</td><td>[:cpa]</td><td>2</td><td></td><td></td><td>0.021978</td><td>0.186813</td><td>0.138504</td><td>0.303091</td><td>0.0682</td><td>0.0796205</td><td>0.021978</td><td>0.203561</td><td>0.25745</td><td>0.17517</td><td>-0.589307</td><td>-0.75654</td></tr><tr><th>10</th><td>8</td><td>[:ptr, :mtr]</td><td>0</td><td></td><td></td><td>0.978022</td><td>NaN</td><td>0.932965</td><td>NaN</td><td>0.006</td><td>0.00700474</td><td>1.0</td><td>1.0</td><td>NaN</td><td>NaN</td><td>NaN</td><td>NaN</td></tr><tr><th>11</th><td>8</td><td>[:ptr, :mtr]</td><td>1</td><td></td><td></td><td>0.021978</td><td>0.978022</td><td>0.064117</td><td>0.932965</td><td>0.006</td><td>0.00700474</td><td>0.021978</td><td>0.0670355</td><td>0.326988</td><td>0.312752</td><td>-0.485468</td><td>-0.5048</td></tr><tr><th>12</th><td>9</td><td>[:mtr]</td><td>0</td><td></td><td></td><td>0.527473</td><td>NaN</td><td>0.377803</td><td>NaN</td><td>0.1</td><td>0.116746</td><td>1.0</td><td>1.0</td><td>NaN</td><td>NaN</td><td>NaN</td><td>NaN</td></tr><tr><th>13</th><td>9</td><td>[:mtr]</td><td>1</td><td></td><td></td><td>0.340659</td><td>0.527473</td><td>0.294297</td><td>0.377803</td><td>0.1</td><td>0.116746</td><td>0.472527</td><td>0.622197</td><td>0.829086</td><td>0.543957</td><td>-0.0814004</td><td>-0.264436</td></tr><tr><th>14</th><td>9</td><td>[:mtr]</td><td>2</td><td></td><td></td><td>0.131868</td><td>0.340659</td><td>0.183922</td><td>0.294297</td><td>0.1</td><td>0.116746</td><td>0.131868</td><td>0.3279</td><td>0.619401</td><td>0.347427</td><td>-0.208028</td><td>-0.459136</td></tr><tr><th>15</th><td>10</td><td>[:ptr]</td><td>0</td><td></td><td></td><td>0.43956</td><td>NaN</td><td>0.377803</td><td>NaN</td><td>0.1</td><td>0.116746</td><td>1.0</td><td>1.0</td><td>NaN</td><td>NaN</td><td>NaN</td><td>NaN</td></tr><tr><th>16</th><td>10</td><td>[:ptr]</td><td>1</td><td></td><td></td><td>0.527473</td><td>0.43956</td><td>0.294297</td><td>0.377803</td><td>0.1</td><td>0.116746</td><td>0.56044</td><td>0.622197</td><td>1.5405</td><td>0.77419</td><td>0.18766</td><td>-0.111153</td></tr><tr><th>17</th><td>10</td><td>[:ptr]</td><td>2</td><td></td><td></td><td>0.032967</td><td>0.527473</td><td>0.183922</td><td>0.294297</td><td>0.1</td><td>0.116746</td><td>0.032967</td><td>0.3279</td><td>0.100007</td><td>0.0560951</td><td>-0.999968</td><td>-1.25108</td></tr><tr><th>18</th><td>11</td><td>[:ugi, :sly, :bvu, :cqu]</td><td>0</td><td></td><td></td><td>0.835165</td><td>NaN</td><td>0.938092</td><td>NaN</td><td>0.00552187</td><td>0.00644655</td><td>1.0</td><td>1.0</td><td>NaN</td><td>NaN</td><td>NaN</td><td>NaN</td></tr><tr><th>19</th><td>11</td><td>[:ugi, :sly, :bvu, :cqu]</td><td>1</td><td></td><td></td><td>0.164835</td><td>0.835165</td><td>0.059424</td><td>0.938092</td><td>0.00552187</td><td>0.00644655</td><td>0.164835</td><td>0.0619082</td><td>3.11574</td><td>2.99072</td><td>0.493561</td><td>0.475775</td></tr><tr><th>20</th><td>12</td><td>[:bvu, :cqu]</td><td>0</td><td></td><td></td><td>1.0</td><td>NaN</td><td>0.493319</td><td>NaN</td><td>0.0682065</td><td>0.0796281</td><td>1.0</td><td>1.0</td><td>NaN</td><td>NaN</td><td>NaN</td><td>NaN</td></tr><tr><th>21</th><td>13</td><td>[:bvu]</td><td>0</td><td></td><td></td><td>1.0</td><td>NaN</td><td>0.625077</td><td>NaN</td><td>0.0433</td><td>0.0505508</td><td>1.0</td><td>1.0</td><td>NaN</td><td>NaN</td><td>NaN</td><td>NaN</td></tr><tr><th>22</th><td>14</td><td>[:cqu]</td><td>0</td><td></td><td></td><td>0.0879121</td><td>NaN</td><td>0.625077</td><td>NaN</td><td>0.0433</td><td>0.0505508</td><td>1.0</td><td>1.0</td><td>NaN</td><td>NaN</td><td>NaN</td><td>NaN</td></tr><tr><th>23</th><td>14</td><td>[:cqu]</td><td>1</td><td>**</td><td>**</td><td>0.912088</td><td>0.0879121</td><td>0.270725</td><td>0.625077</td><td>0.0433</td><td>0.0505508</td><td>0.912088</td><td>0.374923</td><td>23.9548</td><td>17.2973</td><td>1.37939</td><td>1.23798</td></tr><tr><th>24</th><td>15</td><td>[:ugi, :sly]</td><td>0</td><td></td><td></td><td>1.0</td><td>NaN</td><td>0.731798</td><td>NaN</td><td>0.0280065</td><td>0.0326964</td><td>1.0</td><td>1.0</td><td>NaN</td><td>NaN</td><td>NaN</td><td>NaN</td></tr><tr><th>25</th><td>16</td><td>[:ugi]</td><td>0</td><td></td><td></td><td>1.0</td><td>NaN</td><td>0.431809</td><td>NaN</td><td>0.0835</td><td>0.0974826</td><td>1.0</td><td>1.0</td><td>NaN</td><td>NaN</td><td>NaN</td><td>NaN</td></tr><tr><th>26</th><td>17</td><td>[:sly]</td><td>0</td><td></td><td></td><td>0.989011</td><td>NaN</td><td>0.431809</td><td>NaN</td><td>0.0835</td><td>0.0974826</td><td>1.0</td><td>1.0</td><td>NaN</td><td>NaN</td><td>NaN</td><td>NaN</td></tr><tr><th>27</th><td>17</td><td>[:sly]</td><td>1</td><td></td><td></td><td>0.010989</td><td>0.989011</td><td>0.303247</td><td>0.431809</td><td>0.0835</td><td>0.0974826</td><td>0.010989</td><td>0.568191</td><td>0.0158217</td><td>0.00844413</td><td>-1.80075</td><td>-2.07345</td></tr></tbody></table><p>There are some plotting recipes included in the BelugaPlots package. You may want to try out the following:</p><pre><code class="language-julia">using BelugaPlots, Plots, StatsPlots
BelugaPlots.traceplot(chain.trace, burnin=10)
BelugaPlots.traceplot(chain.trace, burnin=10, st=density)
BelugaPlots.bfplot(bayesfactors(chain, burnin=10))
BelugaPlots.ktraceplot(chain.trace, burnin=10)
posteriorE!(chain)
BelugaPlots.eplot(chain.trace)</code></pre><p>To obtain a tree with the marginal posterior mean duplication and loss rate estimates, you can try the following (note this is not likely to give you a very nice image unless your taxon IDs are three letter codes)</p><pre><code class="language-julia">BelugaPlots.doubletree(chain, burnin=10)</code></pre></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="api/">API »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Monday 6 January 2020 08:15">Monday 6 January 2020</span>. Using Julia version 1.2.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
