<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Beluga · Beluga.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">Beluga.jl</span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Beluga</a><ul class="internal"><li><a class="tocitem" href="#Data-preparation-1"><span>Data preparation</span></a></li><li><a class="tocitem" href="#Reversible-jump-MCMC-for-the-inference-of-WGDs-in-a-phylogeny-1"><span>Reversible-jump MCMC for the inference of WGDs in a phylogeny</span></a></li></ul></li><li><a class="tocitem" href="api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Beluga</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Beluga</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/arzwa/Beluga.jl/blob/master/docs/src/index.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Beluga.jl-1"><a class="docs-heading-anchor" href="#Beluga.jl-1">Beluga.jl</a><a class="docs-heading-anchor-permalink" href="#Beluga.jl-1" title="Permalink"></a></h1><p>Beluga is a julia library for the statistical analysis of gene family evolution using phylogenetic birth-death processes. It&#39;s somewhat related to <a href="https://arzwa.github.io/Whale.jl/dev/index.html">Whale.jl</a> as it implements models of duplication loss and whole-genome duplication, but employs gene count data instead of gene trees. The library implements the MCMC sampler of Zwaenepoel &amp; Van de Peer (2019) as well as the reversible-jump MCMC sampler of Zwaenepoel &amp; Van de Peer (2020, <em>in preparation</em>).</p><h2 id="Data-preparation-1"><a class="docs-heading-anchor" href="#Data-preparation-1">Data preparation</a><a class="docs-heading-anchor-permalink" href="#Data-preparation-1" title="Permalink"></a></h2><p>To perform analyses with Beluga, you will need  </p><ol><li>An ultrametric species tree (time tree)</li><li>A <em>phylogenetic profile matrix</em>. If you have a bunch of protein fasta files for a set of species of interest, this can be easily obtained using e.g. <a href="https://github.com/davidemms/OrthoFinder">OrthoFinder</a>.</li></ol><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>If you only need the phylogenetic profile matrix from OrthoFinder, be sure to use the <code>-og</code> flag to stop the OrthoFinder pipeline after orthogroup inference. The phylogenetic profile matrix can be found in the <code>Orthogroups.GeneCount.tsv</code> file.</p></div></div><h2 id="Reversible-jump-MCMC-for-the-inference-of-WGDs-in-a-phylogeny-1"><a class="docs-heading-anchor" href="#Reversible-jump-MCMC-for-the-inference-of-WGDs-in-a-phylogeny-1">Reversible-jump MCMC for the inference of WGDs in a phylogeny</a><a class="docs-heading-anchor-permalink" href="#Reversible-jump-MCMC-for-the-inference-of-WGDs-in-a-phylogeny-1" title="Permalink"></a></h2><p>Load Beluga and required packages:</p><pre><code class="language-">using Beluga, CSV, Distributions</code></pre><p>Or, if you are running a julia session with multiple processes (when you started julia with <code>-p</code> option, or manually added workers using <code>addprocs</code>, see <a href="https://docs.julialang.org/en/v1/manual/parallel-computing/index.html#Parallel-Computing-1">julia docs</a>), run</p><pre><code class="language-julia">using CSV, Distributions
@everywhere using Beluga  # if julia is running in a parallel environment</code></pre><p>Then get some data (for instance from the <code>example</code> directory of the git repo)</p><pre><code class="language-">nw = readline(&quot;../../example/dicots/dicots.nw&quot;)
df = CSV.read(&quot;../../example/dicots/dicots-f01-25.csv&quot;)
model, data = DLWGD(nw, df, 1.0, 1.2, 0.9)</code></pre><p><code>model</code> now refers to the duplication-loss and WGD model (with no WGDs for now), <code>data</code> refers to the phylogenetic profile matrix. The model was initialized with all duplication and loss rates set to 1 and 1.2 respectively. You can check this easily:</p><pre><code class="language-">getrates(model)</code></pre><p>or to get the full parameter vector:</p><pre><code class="language-">asvector(model)</code></pre><p>and you can easily compute likelihoods (and gradients)</p><pre><code class="language-">logpdf!(model, data)</code></pre><p>Now we proceed to the hierarchical model. There is no DSL available (à la Turing.jl, Mamba.jl or Soss.jl) but we use a fairly flexible prior struct. Here is an exaple for the (recommended) independent rates (IR) prior:</p><pre><code class="language-">prior = IRRevJumpPrior(
    # prior covariance matrix (Inverse Wishart prior)
    Σ₀=[1 0. ; 0. 1],

    # multivariate prior on the mean duplication and loss rates
    X₀=MvNormal([0., 0.], [1 0. ; 0. 1]),  

    # prior on the number of WGDs (can be any discrete distribution)
    πK=DiscreteUniform(0,20),

    # prior on the WGD retention rate (should be `Beta`)
    πq=Beta(1,1),

    # prior on η
    πη=Beta(3,1),

    # tree length (determines prior on WGD times)
    Tl=treelength(model))</code></pre><p>We can then construct a chain and run it</p><pre><code class="language-">chain = RevJumpChain(data=data, model=model, prior=prior)
init!(chain)
rjmcmc!(chain, 100, show=10)</code></pre><p>This will log a part of the trace to stdout every <code>show</code> iterations, so that we&#39;re able to monitor a bit whether everything looks sensible. Of course in reality you would sample way longer than n=100 iterations.</p><p>Now we can analyze the trace (in <code>chain.trace</code>), write it to a file, etc. We can also compute Bayes factors to get an idea of the number of WGDs for each branch in the species tree.</p><pre><code class="language-">bayesfactors(chain, burnin=10)</code></pre><p>There are some plotting recipes included in the BelugaPlots package. You may want to try out the following:</p><pre><code class="language-julia">using BelugaPlots, Plots, StatsPlots
BelugaPlots.traceplot(chain.trace, burnin=10)
BelugaPlots.traceplot(chain.trace, burnin=10, st=density)
BelugaPlots.bfplot(bayesfactors(chain, burnin=10))
BelugaPlots.ktraceplot(chain.trace, burnin=10)
posteriorE!(chain)
BelugaPlots.eplot(chain.trace)</code></pre><p>To obtain a tree with the marginal posterior mean duplication and loss rate estimates, you can try the following (note this is not likely to give you a very nice image unless your taxon IDs are three letter codes)</p><pre><code class="language-julia">BelugaPlots.doubletree(chain, burnin=10)</code></pre></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="api/">API »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Sunday 5 January 2020 09:50">Sunday 5 January 2020</span>. Using Julia version 1.2.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
